README.md
# `github-resource-converter` [![NPM version][npm-image]][npm-url]

> <img align="bottom" alt="issue-opened" height="50" width="50"  src="https://cdnjs.cloudflare.com/ajax/libs/octicons/4.4.0/svg/desktop-download.svg"> Convert GitHub Issues and Pull Requests to JSON and CSV from a Terminal or Node.js app.

[![The MIT License][license-image]][license-url]
[![FOSSA Status][fossa-image]][fossa-url]<br>
[![NSP Status][nsp-image]][nsp-url]
[![Dependency Status][daviddm-image]][daviddm-url]
[![Development Dependency Status][daviddm-dev-image]][daviddm-dev-url]<br>
[![MacOS and Ubuntu build statuses][travis-image]][travis-url]
[![Windows build status][appveyor-image]][appveyor-url]
[![Coverage percentage][codacy-coverage-image]][codacy-url]
[![Codacy code quality][codacy-image]][codacy-url]
[![NPMS score][npms-image]][npms-url]
[![NPM downloads per month][npm-downloads-month]][npm-url]

## Table of contents

<!-- ⛔️ AUTO-GENERATED-CONTENT:START (TOC:excludeText=Table of contents) -->
- [1. Overview](#1-overview)
- [2. Installation](#2-installation)
  * [2.1. For Terminal/command-line usage](#21-for-terminalcommand-line-usage)
  * [2.2. As a application dependency](#22-as-a-application-dependency)
- [3. Usage](#3-usage)
  * [3.1. Formatting](#31-formatting)
  * [3.2. Resource types](#32-resource-types)
  * [3.3. Filtering](#33-filtering)
  * [3.4. Exporting](#34-exporting)
    + [3.4.1. Issues](#341-issues)
    + [3.4.2. Pull Requests](#342-pull-requests)
    + [3.4.3. All (open and closed issues and pull requests)](#343-all-open-and-closed-issues-and-pull-requests)
  * [3.5. Command-line flags](#35-command-line-flags)
  * [3.6. Errors](#36-errors)
  * [3.7. Info](#37-info)
- [4. API](#4-api)
  * [4.1. `grc.authenticate({token, type, key}): void`](#41-grcauthenticatetoken-type-key-void)
    + [4.1.1. Parameters](#411-parameters)
    + [4.1.2. Example](#412-example)
  * [4.2. `grc.getAll({owner, repo}): Promise`](#42-grcgetallowner-repo-promise)
  * [4.3. `grc.issues.getForRepo({owner, repo}): Promise`](#43-grcissuesgetforrepoowner-repo-promise)
    + [4.3.1. Parameters](#431-parameters)
    + [4.3.2. Examples](#432-examples)
  * [4.4. `grc.logger`](#44-grclogger)
  * [4.5. `grc.options`](#45-grcoptions)
  * [4.6. `grc.pullRequests.getForRepo({owner, repo}): Promise`](#46-grcpullrequestsgetforrepoowner-repo-promise)
    + [4.6.1. Parameters](#461-parameters)
    + [4.6.2. Examples](#462-examples)
  * [4.7. `grc.save({data, dest}): Promise`](#47-grcsavedata-dest-promise)
    + [4.7.1. Parameters](#471-parameters)
    + [4.7.2. Examples](#472-examples)
  * [4.8. `grc.toCsv({data=[]}): Promise`](#48-grctocsvdata-promise)
    + [4.8.1. Parameters](#481-parameters)
    + [4.8.2. Examples](#482-examples)
- [5. Version](#5-version)
- [6. Contributing](#6-contributing)
- [7. License](#7-license)
<!-- ⛔️ AUTO-GENERATED-CONTENT:END -->

## 1. Overview

`github-resource-converter` (alias `grc`) exports your GitHub and GitHub Enterprise repositories' Issues and Pull Requests to `CSV` and `JSON` file formats. It's helpful whenever you need to:

* **Use spreadsheets** to analyze, modify, print, or summarize large amounts of data with Pivot Tables and other important financial or statistical operations

  _Example:_

  > ```bash
  > # Convert all GitHub issues to CSV:
  > $ github-resource-converter \
  >   --owner foo \
  >   --repo  bar
  > # => Saved "foo-bar-issues-export.2018-03-20T02_11_04_356Z.csv".
  > ```

* **Share data** with other tools like GitLab and JIRA.

  _Example:_

  > ```bash
  > # Convert and save all GitHub Enterprise
  > # Pull Requests as JSON (using the grc alias):
  > $ grc \
  >   --owner gregswindle \
  >   --repo github-resource-converter \
  >   --resource-type pr \
  >   --dest './docs/reports/export.json'
  > # Saved "docs/reports/gregswindle-github-resource-converter-pr-export.2018-03-20T02_18_33_682Z.json".
  > ```

## 2. Installation

1.  **Required:** `github-resource-converter` is written in JavaScript (CommonJS) for [Node.js ![External link][icon-octicon-link-external]](https://nodejs.org/), which must be installed prior to use. Node.js requires **npm**, which is used for installing dependencies. (**npm** installs with Node.js.)

1.  **Recommended:** To avoid rate-limiting, you should [create a personal access token ![External link][icon-octicon-link-external]](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/) and save your personal access token.

    * **MacOS and Unix**

      ```bash
      $ mkdir -p /usr/local/etc/github-resource-center/envvars/
      $ touch /usr/local/etc/github-resource-center/envvars/.env
      $ echo "GITHUB_ACCESS_TOKEN="{your-personal-access-token-value}" > \
        /usr/local/etc/github-resource-center/envvars/.env
      ```

    * **Windows**

      ```shell
      > md -p C:\usr\local\etc\github-resource-center\envvars\
      > touch C:\usr\local\etc\github-resource-center\envvars\.env
      > echo "GITHUB_ACCESS_TOKEN="{your-personal-access-token-value}" >>
        C:\usr\local\etc\github-resource-center\envvars\.env
      ```

### 2.1. For Terminal/command-line usage

```bash
# Install globally to execute from a Terminal/command-line
$ npm i -g github-resource-converter
```

### 2.2. As a application dependency

```bash
# Install as a dependency within a Node.js app
$ npm i --save github-resource-converter
```

## 3. Usage

> <img align="left" alt="terminal" height="30" width="30" src="https://cdnjs.cloudflare.com/ajax/libs/octicons/4.4.0/svg/terminal.svg"> The following examples assume that `github-resource-converter` is installed globally and invoked from a Terminal (command-line interface)

### 3.1. Formatting

`--dest export.[csv|json]`

> You can convert GitHub (Enterprise) Issues and Pull Requests into two file formats: `CSV` and `JSON`.

* CSV is the default format.
* JSON formatting requires a `--dest` value with a `.json` file extension.

### 3.2. Resource types

`--resource-type, -t`

> Convert and export GitHub (Enterprise) by `--resource-type`: `issues`, `pull_requests`, or both (`all`).

* `issues` is the default `resource-type`.
* `prs` require a `--resource-type` or `-t` value of

  * `pr`
  * `prs`
  * `pull_request`
  * `pull_requests`

* `--resource-type all` will export all `issues` and `prs` into a single file.

### 3.3. Filtering

> ![alert][icon-octicon-alert] **Filtering is currently unavailable.**
>
> If you're interested in [CONTRIBUTING](#contributing) to features like filters--e.g., only select
> "open" issues--we're happily accepting pull requests!

### 3.4. Exporting

#### 3.4.1. Issues

* **CSV**

  _GitHub:_

  > ```bash
  > # GET https://api.github.com/rrepos/:owner/:repo/issues
  > $ github-resource-converter \
  >   --owner gregswindle \
  >   --repo  github-resource-converter \
  >   --dest  './docs/reports/export.csv'
  > ```

  _GitHub Enterprise:_

  > ```bash
  > # GET https://api.ecorp.com/api/v3/repos/:owner/:repo/issues
  > $ grc --base-url https://api.ecorp.com/api/v3 \
  >   --owner evilcorp \
  >   --repo ecoin
  > ```

* **JSON**

  _GitHub:_

  > ```bash
  > # GET https://api.github.com/repos/:owner/:repo/issues
  > $ github-resource-converter \
  >   --owner gregswindle \
  >   --repo  github-resource-converter \
  >   --dest  './docs/reports/export.json'
  > ```

  _GitHub Enterprise:_

  > ```bash
  > # GET https://api.ecorp.com/api/v3/repos/:owner/:repo/issues
  > $ grc --base-url https://api.ecorp.com/api/v3 \
  >   --owner evilcorp \
  >   --repo ecoin
  >   --dest ./export.json
  > ```

#### 3.4.2. Pull Requests

* **CSV**

  _GitHub:_

  > ```bash
  > # GET https://api.github.com/repos/:owner/:repo/pulls
  > $ github-resource-converter \
  >   --owner gregswindle \
  >   --repo  github-resource-converter \
  >   --resource-type pr
  > ```

  _GitHub Enterprise:_

  > ```bash
  > # GET https://api.ecorp.com/api/v3/repos/:owner/:repo/pulls
  > $ grc --base-url https://api.ecorp.com/api/v3 \
  >   --owner evilcorp \
  >   --repo ecoin \
  >   --resource-type pr
  > ```

* **JSON**

  _GitHub:_

  > ```bash
  > # GET https://api.github.com/repos/:owner/:repo/pulls
  > $ github-resource-converter \
  >   --owner gregswindle \
  >   --repo  github-resource-converter \
  >   --resource-type pr
  >   --dest './export.json'
  > ```

  _GitHub Enterprise:_

  > ```bash
  > # GET https://api.ecorp.com/api/v3/repos/:owner/:repo/pulls
  > $ grc --base-url https://api.ecorp.com/api/v3 \
  >   --owner evilcorp \
  >   --repo ecoin \
  >   --resource-type pr
  >   --dest './export.json'
  > ```

#### 3.4.3. All (open and closed issues and pull requests)

* **CSV**

  _GitHub:_

  > ```bash
  > # GET https://api.github.com/repos/:owner/:repo/pulls
  > $ github-resource-converter \
  >   --owner gregswindle \
  >   --repo  github-resource-converter \
  >   --resource-type all
  > ```

* _GitHub Enterprise:_

  > ```bash
  > # GET https://api.ecorp.com/api/v3/repos/:owner/:repo/pulls
  > $ grc --base-url https://api.ecorp.com/api/v3 \
  >   --owner evilcorp \
  >   --repo ecoin \
  >   --resource-type all
  > ```

* **JSON**

  _GitHub:_

  > ```bash
  > # GET https://api.github.com/repos/:owner/:repo/pulls
  > $ github-resource-converter \
  >   --owner gregswindle \
  >   --repo  github-resource-converter \
  >   --resource-type all
  >   --dest './export.json'
  > ```

* _GitHub Enterprise:_

  > ```bash
  > # GET https://api.ecorp.com/api/v3/repos/:owner/:repo/pulls
  > $ grc --base-url https://api.ecorp.com/api/v3 \
  >   --owner evilcorp \
  >   --repo ecoin \
  >   --resource-type all
  >   --dest './export.json'
  > ```

### 3.5. Command-line flags

<dl>
  <dt><code>--owner, -o</code></dt>
  <dd><p><strong>Required.</strong> The GitHub account name or organization name.</p></dd>
  <dt><code>--repo, -r</code></dt>
  <dd><p><strong>Required.</strong> The name of the GitHub (or GitHub Enterprise) repository.</p></dd>
  <dt><code>--base-url</code></dt>
  <dd><p>The GitHub REST API v3 URL origin, or a GitHub Enterprise URL origin and path-prefix.</p>
    <table>
     <tbody><tr><th>Default value:</th><td><code>https://api.github.com</code></td></tr></tbody>
    </table>
  </dd>
  <dt><code>--dest, -d</code></dt>
  <dd><p>The destination path and file name of the CSV.</p>
    <table>
     <tbody><tr><th>Default value:</th><td><samp>./export.csv</samp></td></tr></tbody>
    </table>
  </dd>
  <dt><code>--no-auto-filename</code></dt>
  <dd><p>Disable automatic file naming.</p>
    <table>
     <tbody><tr><th>Default value:</th><td><samp>false</samp></td></tr></tbody>
    </table>
  </dd>
  <dt><code>--resource-type, -t</code></dt>
  <dd><p>Declares whether to convert and export Issues, Pull Requests, or All.</p>
    <table>
      <tr>
        <th>Default&nbsp;value:</th>
        <td><samp>issues</samp></td>
      </tr>
      <tr>
        <th>Valid values:</th>
        <td><samp>all</samp>, <samp>issue</samp>, <samp>issues</samp>, <samp>pr</samp>, <samp>prs</samp>, <samp>pull_request</samp>, <samp>pull_requests</samp></td>
      </tr>
    </table>
  </dd>
</dl>

### 3.6. Errors

Errors are written to the console (`stdout`) as JSON:

```bash
# Attempt to fetch issues from a repository that doesn't exist
$ grc --owner example --repo error
[2018-03-20T02:31:24.737Z] ERROR: github-resource-converter/48219 on User.router.home: {"message":"Not Found","documentation_url":"https://developer.github.com/v3/issues/#list-issues-for-a-repository"}
  HttpError: {"message":"Not Found","documentation_url":"https://developer.github.com/v3/issues/#list-issues-for-a-repository"}
      at response.text.then.message (/p/a/t/h/github-resource-converter/node_modules/@octokit/rest/lib/request/request.js:56:19)
      at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)
```

### 3.7. Info

The `--help` flag displays all options:

```shell
$ grc --help

  Convert and export GitHub resources--Issues and Pull Requests--to CSV and JSON.

  Usage

    $ grc [options] [info]
    $ github-resource-converter [options] [info]

  Options

    --base-url           The GitHub REST API v3 URL origin, or a
                         GitHub Enterprise URL origin and path-prefix.
                         [Default: 'https://api.github.com']
    --dest,          -d  The CSV's destination path and file name.
                         [Default: './resources.csv']
    --no-auto-filename   Disable automatic file naming.
                         [Default: false]
    --owner,         -o  The GitHub account name or organization name.
    --repo,          -r  The name of the GitHub (or GitHub enterprise)
                         repository.
    --resource-type, -t  "issues", "prs", or "all".
                         [Default: 'issues']

  Info

    --help     Show this dialog.
    --version  Display the installed semantic version.

  Examples

    $ grc --owner github --repo hub
      // => Exported CSV to /path/of/cwd/issues.csv.

    $ grc --owner github --repo hub -dest './reports/issues/YYYY-MM-DD.csv'
      // => Exported CSV to /path/to/reports/issues/YYYY-MM-DD.csv.

    $ grc --owner example --repo error
      // =>
      [2018-03-19T08:04:06.596Z] ERROR: github-resource-converter/00000 on localhost: Cannot destructure property `data` of 'undefined' or 'null'.
        TypeError: Cannot destructure property `data` of 'undefined' or 'null'.
            at paginate (/p/a/t/h/github-resource-converter/lib/base-resource-converter.js:39:16)
            at <anonymous>
            at process._tickCallback (internal/process/next_tick.js:188:7
```

Use the `--version` flag to see which version you have installed:

```bash
$ github-resource-converter --version
# => 1.0.1
```

## 4. API

> [![beaker][icon-octicon-beaker] Test the `github-resource-converter (grc)` API in your Web browser ![link-external][icon-octicon-link-external]][runkit-grc-url].

### 4.1. `grc.authenticate({token, type, key}): void`

> ![Info][icon-octicon-info] Most GitHub API calls don't require authentication. Rules of thumb:
>
> 1.  If you can see the information by visiting the site without being logged in, you don't have to be authenticated to retrieve the same information through the API.
> 1.  If you want to change data, you have to be authenticated.
>
> octokit/rest.js. (2018). GitHub. Retrieved 21 March 2018, from <https://github.com/octokit/rest.js#authentication> ![link-external][icon-octicon-link-external]

#### 4.1.1. Parameters

| Name  | Type   | Description                                                      | Notes |
| :---- | :----- | :--------------------------------------------------------------- | :---- |
| key   | String |                                                                  |       |
| token | String |                                                                  |       |
| type  | Enum   | `basic`, `oauth`, `oauth-key-secret`, `token`, and `integration` |       |

#### 4.1.2. Example

```javascript
// Token (https://github.com/settings/tokens)
grc.authenticate({
  token: 'secrettoken123',
  type: 'token'
})
```

### 4.2. `grc.getAll({owner, repo}): Promise`

Retrieve all open and closed Issues and Pull Requests from a GitHub project.

![GET][rest-get-img]

```http
/repos/:owner/:repo/issues
/repos/:owner/:repo/pulls
```

> **![info][icon-octicon-info] `grc.getAll` combines the results of `grc.issues.getForRepo` and `grc.pullRequests.getForRepo` into a single `Array<object>`.**
>
> For more information, see
>
> 1.  [`grc.issues.getForRepo`](#43-grcissuesgetforrepoowner-repo-promise) and
> 1.  [`grc.pullRequests.getForRepo`](#46-grcpullrequestsgetforrepoowner-repo-promise) below.

### 4.3. `grc.issues.getForRepo({owner, repo}): Promise`

Retrieve all open and closed **issues** for a repository with this proxy method for [`octokit.issues.getForRepo` ![link-external][icon-octicon-link-external]](https://octokit.github.io/rest.js/#api-Issues-getForRepo).

![GET][rest-get-img]

```http
/repos/:owner/:repo/issues
```

#### 4.3.1. Parameters

<table>
  <thead>
    <tr>
      <th style="width: 30%">Field</th>
      <th style="width: 10%">Type</th>
      <th style="width: 60%">Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><samp>owner</samp></td>
      <td>
        string
      </td>
      <td></td>
    </tr>
    <tr>
      <td><samp>repo</samp></td>
      <td>
        string
      </td>
      <td></td>
    </tr>
    <tr>
      <td><samp>milestone</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        string
      </td>
      <td></td>
    </tr>
    <tr>
      <td><samp>state</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        string
      </td>
      <td>
        <p>open, closed, or all</p>
        <p>Default value: <code>open</code></p>
        <p>Allowed values:
          <code>open</code>, 
          <code>closed</code>, 
          <code>all</code>
        </p>
      </td>
    </tr>
    <tr>
      <td><samp>assignee</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        string
      </td>
      <td>
        <p>String User login, <code>none</code> for Issues with no assigned User. <code>*</code> for Issues with any assigned User.</p>
      </td>
    </tr>
    <tr>
      <td><samp>creator</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        string
      </td>
      <td>
        <p>The user that created the issue.</p>
      </td>
    </tr>
    <tr>
      <td><samp>per_page</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        number
      </td>
      <td>
        <p>A custom page size up to 100. Default is 30.</p>
        <p>Default value: <code>30</code></p>
      </td>
    </tr>
    <tr>
      <td><samp>labels</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        string
      </td>
      <td>
        <p>String list of comma separated Label names. Example: bug,ui,@high</p>
      </td>
    </tr>
    <tr>
      <td><samp>sort</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        string
      </td>
      <td>
        <p>Default value: <code>created</code></p>
        <p>Allowed values:
          <code>created</code>,
          <code>updated</code>,
          <code>comments</code>
        </p>
      </td>
    </tr>
    <tr>
      <td><samp>direction</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        string
      </td>
      <td>
        <p>Default value: <code>desc</code></p>
        <p>Allowed values:
          <code>asc</code>, 
          <code>desc</code>
        </p>
      </td>
    </tr>
    <tr>
      <td><samp>since</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        date
      </td>
      <td>
        <p>Timestamp in ISO 8601 format: YYYY-MM-DDTHH:MM:SSZ</p>
      </td>
    </tr>
    <tr>
      <td><samp>page</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        number
      </td>
      <td>
        <p>Page number of the results to fetch.</p>
      </td>
    </tr>
    <tr>
      <td><samp>mentioned</samp> <img align="right" alt="optional" height="22" width="60" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16"></td>
      <td>
        string
      </td>
      <td>
        <p>String User login.</p>
      </td>
    </tr>
  </tbody>
</table>

#### 4.3.2. Examples

* _**async/await**:_

  > ```javascript
  > const result = await grc.issues.getForRepo({
  >   owner: 'gregswindle',
  >   repo: 'eslint-plugin-crc'
  > })
  > ```

* _**Promise**:_

  > ```javascript
  > grc.issues
  >   .getForRepo({
  >     owner: 'gregswindle',
  >     repo: 'eslint-plugin-crc'
  >   })
  >   .then(result => {})
  >   .catch(err => {})
  > ```

### 4.4. `grc.logger`

A [`trentm/node-bunyan` ![link-external][icon-octicon-link-external]][node-bunyan-url] logger instance that uses the `LONG` [`thlorenz/bunyan-format` ![link-external][icon-octicon-link-external]][bunyan-format-url] writeable
stream for output.

> ![info][icon-octicon-info] **`bunyan.INFO`** is the default log LEVEL.

```javascript
Logger {
  domain: null,
  _events: {},
  _eventsCount: 0,
  _maxListeners: undefined,
  _level: 30,
  streams:
    [{
      type: 'stream',
      stream: BunyanFormatWritable {
      },
      closeOnExit: false,
      level: 30,
      raw: false
    }],
  serializers: null,
  src: false,
  fields: {
    name: 'github-resource-converter',
     hostname: 'Gregorys-MBP-2.fios-router.home',
    pid: 76698
  },

  debug(): void,
  error(): void,
  fatal(): void,
  info(): void,
  trace(): void,
  warn(): void
}
```

### 4.5. `grc.options`

Contains default values for `api`, `cli`, and `meta`data.

---

<details>
  <summary><strong>Toggle view of <samp>grc.options</samp></strong>.</summary>
  <p>

```javascript
{
  "api": {
    "baseUrl": "https://api.github.com",
    "headers": {
      "Accept":
        "application/vnd.github.v3+json, application/vnd.github.symmetra-preview+json",
      "user-agent": "gregswindle/github-resource-converter v1.0.1"
    },
    "owner": null,
    "repo": null,
    "requestMedia": "application/vnd.github.v3+json",
    "timeout": 0
  },
  "cli": {
    "flags": {
      "baseUrl": {
        "default": "https://api.github.com",
        "type": "string"
      },
      "dest": {
        "alias": "d",
        "default": "./export.csv",
        "type": "string"
      },
      "no-auto-filename": {
        "default": false,
        "type": "boolean"
      },
      "owner": {
        "alias": "o",
        "type": "string"
      },
      "repo": {
        "alias": "r",
        "type": "string"
      },
      "resource-type": {
        "alias": "t",
        "default": "issues",
        "type": "string"
      }
    }
  },
  "meta": {
    "input": [],
    "flags": {
      "noAutoFilename": false,
      "baseUrl": "https://api.github.com",
      "dest": "./export.csv",
      "d": "./export.csv",
      "resourceType": "issues",
      "t": "issues"
    },
    "pkg": {
      "name": "github-resource-converter",
      "description":
        "Convert and export GitHub resources--Issues and Pull Requests--to CSV and JSON.",
      "version": "1.0.1",
      "author": {
        "name": "Greg Swindle",
        "email": "greg@swindle.net",
        "url": "https://github.com/gregswindle"
      },
      "bin": {
        "grc": "lib/cli.js"
      },
      "bugs": {
        "url": "https://github.com/gregswindle/github-resource-converter/issues"
      },
      "commitplease": {
        "style": "angular",
        "types": [
          "build",
          "chore",
          "ci",
          "docs",
          "feat",
          "fix",
          "perf",
          "refactor",
          "revert",
          "style",
          "test"
        ],
        "scope": "\\S+.*"
      },
      "contributors": [],
      "dependencies": {
        "@octokit/rest": "15.2.6",
        "bunyan": "1.8.12",
        "bunyan-format": "0.2.1",
        "dotenv": "5.0.1",
        "fs-extra": "5.0.0",
        "insight": "0.10.1",
        "jsonexport": "2.0.11",
        "lodash": "4.17.5",
        "meow": "4.0.0"
      },
      "devDependencies": {
        "@semantic-release/changelog": "2.0.1",
        "@semantic-release/git": "4.0.1",
        "@semantic-release/npm": "3.2.4",
        "ajv": "6.4.0",
        "ajv-keywords": "3.1.0",
        "codacy-coverage": "2.1.1",
        "commitplease": "3.2.0",
        "coveralls": "3.0.0",
        "eslint": "4.19.1",
        "eslint-config-prettier": "^2.4.0",
        "eslint-config-scanjs": "1.0.0-beta4",
        "eslint-config-standard": "11.0.0",
        "eslint-config-xo": "0.20.1",
        "eslint-plugin-import": "2.10.0",
        "eslint-plugin-jsdoc": "3.5.0",
        "eslint-plugin-json": "1.2.0",
        "eslint-plugin-no-unsafe-innerhtml": "1.0.16",
        "eslint-plugin-node": "6.0.1",
        "eslint-plugin-prettier": "^2.2.0",
        "eslint-plugin-promise": "3.7.0",
        "eslint-plugin-security": "1.4.0",
        "eslint-plugin-standard": "3.0.1",
        "eslint-plugin-unicorn": "4.0.2",
        "eslint-plugin-xss": "0.1.9",
        "fixpack": "2.3.1",
        "husky": "^0.14.3",
        "jest": "22.4.3",
        "jest-runner-eslint": "0.4.0",
        "lec": "^1.0.1",
        "lint-staged": "7.0.3",
        "markdown-magic": "0.1.21",
        "markdown-magic-dependency-table": "1.3.2",
        "markdown-magic-install-command": "1.3.1",
        "markdown-magic-package-scripts": "1.2.1",
        "nsp": "^3.2.1",
        "prettier": "1.11.1",
        "semantic-release": "15.1.5"
      },
      "engines": {
        "node": ">=8.0.0"
      },
      "eslintIgnore": ["lib/__tests__/coverage/**"],
      "files": ["lib"],
      "homepage":
        "https://github.com/gregswindle/github-resource-converter/#readme",
      "jest": {
        "automock": false,
        "collectCoverage": true,
        "coverageDirectory": "lib/__tests__/coverage",
        "coverageThreshold": {
          "global": {
            "branches": 100,
            "functions": 100,
            "lines": 100,
            "statements": 100
          }
        },
        "moduleDirectories": ["node_modules", "lib"],
        "testMatch": [
          "<rootDir>/lib/__tests__/**/*.test.js",
          "<rootDir>/lib/__tests__/*.test.js"
        ],
        "watchman": false
      },
      "jest-runner-eslint": {
        "cliOptions": {
          "config": ".github/config/jest-runner-eslint.config.json",
          "fix": true
        }
      },
      "keywords": [
        "QA",
        "convert",
        "converter",
        "converter",
        "csv",
        "export",
        "github",
        "google sheets",
        "issue",
        "json",
        "pr",
        "pull request",
        "pull-request",
        "quality assurance",
        "quality-assurance",
        "reporting",
        "reports",
        "save",
        "sheets",
        "spreadsheets",
        "testing"
      ],
      "license": "MIT",
      "lint-staged": {
        "*.js": ["npm run lint:js", "git add"],
        "*.json": ["npm run lint:json", "git add"],
        "*.md": ["npm run lint:md", "npm run docs", "git add"]
      },
      "main": "lib/index.js",
      "prettier": {
        "semi": false,
        "singleQuote": true
      },
      "release": {
        "generateNotes": {
          "preset": "angular"
        },
        "prepare": [
          "@semantic-release/changelog",
          "@semantic-release/npm",
          "@semantic-release/git"
        ],
        "verifyConditions": ["@semantic-release/npm", "@semantic-release/git"]
      },
      "repository": {
        "type": "git",
        "url":
          "git+https://github.com/gregswindle/github-resource-converter.git"
      },
      "scripts": {
        "docs": "npm run docs:toc",
        "docs:toc":
          "md-magic --config '.github/config/markdown.config.js' --path '**/*.md' --ignore 'node_modules'",
        "lint":
          "npm run lint:js && npm run lint:json && npm run lint:manifest && npm run lint:md",
        "lint:js":
          "node_modules/.bin/eslint -c .github/config/.eslintrc.yml --ext .js . --fix",
        "lint:json":
          "prettier ./**/*.json --ignore-path '.github/config/.prettierignore' --write",
        "lint:manifest": "fixpack",
        "lint:md": "prettier ./**/*.md -prose-wrap always --write",
        "posttest:ci:coverage:codacy":
          "cat ./lib/__tests__/coverage/lcov.info | codacy-coverage",
        "precommit": "lint-staged",
        "prepare": "npm run lint && npm run security",
        "prepublish":
          "lec lib/cli.js -c LF && npm run security && semantic-release -d",
        "prepublishOnly": "npm run prepare",
        "pretest": "npm run lint",
        "security": "npm run security:nsp:scan",
        "security:nsp:scan": "nsp check",
        "test": "jest",
        "test:config": "jest --showConfig",
        "test:watch": "jest ./lib/__tests__/*.test.js --watch",
        "test:watch:all": "jest ./lib/__tests__/*.test.js --watchAll"
      },
      "readme": "ERROR: No README data found!",
      "_id": "github-resource-converter@1.0.1"
    },
    "help":

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.1.0"></a>

# [1.1.0](https://github.com/gregswindle/github-resource-converter/compare/v1.0.1...v1.1.0) (2018-04-09)

### Features

* **analytics:** add yeomon/insight ([c4588c2](https://github.com/gregswindle/github-resource-converter/commit/c4588c2)), closes [#20](https://github.com/gregswindle/github-resource-converter/issues/20)

<a name="1.0.1"></a>

## [1.0.1](https://github.com/gregswindle/github-resource-converter/compare/v1.0.0...v1.0.1) (2018-04-04)

### Bug Fixes

* **save:** add save method to module ([7704a9f](https://github.com/gregswindle/github-resource-converter/commit/7704a9f)), closes [#18](https://github.com/gregswindle/github-resource-converter/issues/18)
* **save:** ensure module's save method functions ([1f7a111](https://github.com/gregswindle/github-resource-converter/commit/1f7a111)), closes [#16](https://github.com/gregswindle/github-resource-converter/issues/16)
* **save:** implement "save" method for Node.js API ([#17](https://github.com/gregswindle/github-resource-converter/issues/17)) ([7e22c30](https://github.com/gregswindle/github-resource-converter/commit/7e22c30))

<a name="1.0.0"></a>

## 1.0.0 (2018-04-03)

### Bug Fixes

* **cli:** import fs-extra ([#11](https://github.com/gregswindle/github-resource-converter/issues/11)) ([3b988cf](https://github.com/gregswindle/github-resource-converter/commit/3b988cf)), closes [#1](https://github.com/gregswindle/github-resource-converter/issues/1) [#4](https://github.com/gregswindle/github-resource-converter/issues/4) [/travis-ci.org/gregswindle/github-resource-converter/jobs/355791728#L2892](https://github.com//travis-ci.org/gregswindle/github-resource-converter/jobs/355791728/issues/L2892) [#1](https://github.com/gregswindle/github-resource-converter/issues/1) [#2](https://github.com/gregswindle/github-resource-converter/issues/2) [#3](https://github.com/gregswindle/github-resource-converter/issues/3) [#4](https://github.com/gregswindle/github-resource-converter/issues/4) [#10](https://github.com/gregswindle/github-resource-converter/issues/10)

### Features

* **deployment:** enable semantic-release ([#13](https://github.com/gregswindle/github-resource-converter/issues/13)) ([6b85819](https://github.com/gregswindle/github-resource-converter/commit/6b85819)), closes [#12](https://github.com/gregswindle/github-resource-converter/issues/12)


lib/templates/help/en-us.js
module.exports = {
  'help': `
Usage

  $ grc [options] [info]
  $ github-resource-converter [options] [info]

Options
  --base-url           The GitHub REST API v3 URL origin, or a
                       GitHub Enterprise URL origin and path-prefix.
                       [Default: 'https://api.github.com']
  --dest,          -d  The CSV's destination path and file name.
                       [Default: './resources.csv']
  --no-auto-filename       Don't append an ISO 8601-like timestamp to the
                       output file.
                       [Default: false]
  --owner,         -o  The GitHub account name or organization name.
  --repo,          -r  The name of the GitHub (or GitHub enterprise)
                       repository.
  --resource-type, -t  "issues", "prs", or "all".
                       [Default: 'issues']

Info

  --help     Show this dialog.
  --version  Display the installed semantic version.

Examples

  $ grc --owner github --repo hub
    // => Exported CSV to /path/of/cwd/issues.csv.

  $ grc --owner github --repo hub -dest './reports/issues/YYYY-MM-DD.csv'
    // => Exported CSV to /path/to/reports/issues/YYYY-MM-DD.csv.

  $ grc --owner example --repo error
    // =>
    [2018-03-19T08:04:06.596Z] ERROR: github-resource-converter/00000 on localhost: Cannot destructure property \`data\` of 'undefined' or 'null'.
      TypeError: Cannot destructure property \`data\` of 'undefined' or 'null'.
          at paginate (/p/a/t/h/github-resource-converter/lib/base-resource-converter.js:39:16)
          at <anonymous>
          at process._tickCallback (internal/process/next_tick.js:188:7)
`
}
